### ISSUE TITLE
Fine-tune Facenet model

### ISSUE THREAD

**Issue by mustafak1** [serengil/deepface#687]
Can I fine-tune the Facenet model on my own dataset? I have a dataset of 20,000 individuals and about 100,000 augmented images. I want to generate better embeddings for my custom kind of database (including obstructions such as facial hair, glasses etc.)

**Comment by serengil** [comment]
Yes, you can. Go for it!

**Comment by mustafak1** [comment]
I tried using the triplet loss function. However when i train, regardless of the epochs or the number of steps per epoch, the accuracy never goes above 0.009. Any suggestions/help?

**Comment by serengil** [comment]
Did you normalize the input?

**Comment by mustafak1** [comment]
Yes I have

**Comment by mustafak1** [comment]
If it provides any insight this is my loss function: 

`def triplet_loss(y_true, y_pred, alpha=ALPHA):
    anchor, positive, negative = y_pred[0], y_pred[1], y_pred[2]
    pos_dist = tf.reduce_sum(tf.square(tf.subtract(anchor, positive)), axis=-1)
    neg_dist = tf.reduce_sum(tf.square(tf.subtract(anchor, negative)), axis=-1)
    basic_loss = tf.add(tf.subtract(pos_dist, neg_dist), alpha)
    loss = tf.reduce_sum(tf.maximum(basic_loss, 0.0))
    return loss`

And this is training script:

`model_name = 'Facenet'
model = DeepFace.build_model(model_name)
print("Built model ", model_name)

num_layers = len(model.layers)
print('Number of layers:', num_layers)

for layer in model.layers[:-105]:
    layer.trainable  =  False
   

model.summary()

input_shape=(IMAGE_SIZE, IMAGE_SIZE,3)
A = Input(shape=input_shape, name = 'anchor')
P = Input(shape=input_shape, name = 'anchorPositive')
N = Input(shape=input_shape, name = 'anchorNegative')

enc_A = model(A)
enc_P = model(P)
enc_N = model(N)


# Callbacks
early_stopping = EarlyStopping(monitor='accuracy', patience=10, min_delta=0.00005)
STAMP = 'facenet_%d'%(len(paths)) 
checkpoint_dir = './' + 'checkpoints/' + str(int(time.time())) + '/'

if not os.path.exists(checkpoint_dir):
    os.makedirs(checkpoint_dir)

bst_model_path = checkpoint_dir + STAMP + '.h5'
tensorboard = TensorBoard(log_dir=checkpoint_dir + "logs/{}".format(time.time()))

# Model
lr = 0.0001 # define learning rate
optimize = Adam(lr=lr)
tripletModel = Model(inputs=[A, P, N], outputs=[enc_A, enc_P, enc_N])
tripletModel.compile(optimizer = optimize, loss = triplet_loss, metrics=['accuracy'])

gen = batch_generator(BATCH_SIZE)
tripletModel.fit(gen, epochs=NUM_EPOCHS,steps_per_epoch=STEPS_PER_EPOCH,callbacks=[early_stopping, tensorboard])`

**Comment by serengil** [comment]
I am not train any model in this project for facial recognition. They are all pre-trained. You should ask this question to the reference model repo.

**Comment by tigflanker** [comment]
@serengil Hi serengil, deepface framework is very convenience first, thank you for provide such a useful system.
I found all pretrain models didn't have a good performance in Asian face data.
So, please tell me where the pretrain models you collected if convenience.
Asian face data is sensitive to collect, i will find if more better performance pretrain model have.
